### PR TITLE
PLT-360 ADR documenting our decision to commit the cardano-api package

### PR DESCRIPTION
**Pull request by koslambrou**
See https://plutus-apps--734.org.readthedocs.build/en/734/adr/0012-commit-to-datatypes-in-cardano-api.html

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [x] Reviewer requested

### CODE REVIEW

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+emulator and the chain indexer.
+
+Decision
+--------
+
+* We will remove our datatype representation of a Cardano transaction
+  (``Ledger.Tx.Internal.Tx`` in ``plutus-ledger``) and fully commit to
+  ``Cardano.Api.Tx.Tx era`` in the codebase.
+
+* We will replace any use of ``plutus-ledger-api`` types by ``cardano-api`` types
```
Comment (berewt) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:64]: Maybe should we move this after the introduction of `cardano-api-extended` and talk about "`cardano-api-extended` types" here, as we may encounter scenarios where we need a slightly different representation.

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    * a simplified and working transaction balancing function (mainly the
+      ``Ledger.Fee.makeAutoBalancedTransaction`` in ``plutus-apps``)
+
+    * validation rules (most of what's in the current ``Ledger.Validation``)
+
+* We will improve ``cardano-api`` through ``cardano-api-extended`` and regularly
+  push changes upstream when possible.
+
+* We will have packages in ``plutus-apps`` depend on ``cardano-api-extended`` which
+  will re-export the modules from the hydra-cardano-api_ package.
```
Comment (berewt) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:91]: Not sure of the meaning here, will packages `plutus-apps` re-export the modules, or will `cardano-api-extended` do it? I think it's the latter, but then, why don't we mention the re-export policy in the paragraph dedicated to the creation of `hydra-cardano-api`?

**Review by berewt — APPROVED** [review]
Some minor comments, no blocker.

**Review by sjoerdvisscher — APPROVED** [review]

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+Argument
+--------
+
+The ``cardano-api`` package is expected to be the supported entrypoint for
+clients to interact with Cardano chain in Haskell in the foreseable future.
+Therefore, we should extensively use this package and upstream changes as much
+as possible so that users not using the packages in ``plutus-apps`` can still
+have a good experience writing Plutus application using ``cardano-api``.
+
+Alternatives
```
Comment (raduom) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:112]: Isn't using `hydra-cardano-api` also an alternative?

**Review by raduom — APPROVED** [review]
LGTM

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst (hunk 2 above)
Comment (koslambrou) [reply doc/adr/0012-commit-to-datatypes-in-cardano-api.rst]: :+1: Moved the re-export policy in the paragraph dedicated to the creation of hydra-cardano-api

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+Draft
+
+Context
+-------
+
+Since the genesis of the ``plutus-apps`` repository, the components have been
+historically using the types in the ``plutus-ledger-api`` package (which is now
+part of ``plutus`` repository) in the off-chain part of a Plutus application.
+
+This was desirable in order to start designing a way to build Plutus application before the ``cardano-ledger`` actually supported the Alonzo-era features.
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:25]: suggest plural "applications"

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+Of course, this resulted in the unintented consequence that we used ``TxInfo`` types (types that are designed to be used in Plutus scripts) in off-chain code.
+This wouldn't has been a problem if there was 1:1 relationship between on-chain and off-chain types.
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:27]: "has been" should be "have been";
suggest inserting "a" --> if there was a 1:1 relationship...

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+Let's take the example of the ``TxOut`` representation of ``plutus-ledger-api`` for ``PlutusV2``.
+
+.. code-block:: haskell
+
+  data TxOut = TxOut {
+    ...
+    txOutReferenceScript :: Maybe ScriptHash
+    }
+
+As we can see, the ``TxOut`` can optionnaly store the ``ScriptHash`` of the referenced script.
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:39]: typo: optionnaly should be "optionally"

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+This proved that we need to start moving away from `plutus-ledger-api` types in
+the off-chain part of Plutus applications, especially in components like the
+emulator and the chain indexer.
+
+Decision
+--------
+
+* We will create a ``cardano-api-extended`` cabal project which will contain
+  features and utilies on top of the ``cardano-api`` package.
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:61]: typo: "utilies" should be "utilities"

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+    * type synonyms for working with the latest era
+
+    * a simplified and working transaction balancing function (mainly the
+      ``Ledger.Fee.makeAutoBalancedTransaction`` in ``plutus-apps``)
+
+    * validation rules (most of what's in the current ``Ledger.Validation``)
+
+  The ``cardano-api-extended`` package will re-export the modules from the
+  hydra-cardano-api_ package which contain type synonyms for working with the
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:73]: "contain" should be "contains"

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+* We will modify the ``plutus-contract`` emulator to fully use the
+  ``cardano-ledger`` transaction validation rules and we will remove our custom
+  validation rules (module ``Ledger.Index`` in ``plutus-ledger``).
+
+Argument
+--------
+
+The ``cardano-api`` package is expected to be the supported entrypoint for
+clients to interact with Cardano chain in Haskell in the foreseable future.
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:109]: "foreseable" should be "foreseeable"

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+Therefore, we should extensively use this package and upstream changes as much
+as possible so that users not using the packages in ``plutus-apps`` can still
+have a good experience writing Plutus application using ``cardano-api``.
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:112]: suggest plural "applications"

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This is currently what the ``plutus-apps`` repository is partially doing.
+The main problem is that this requires significant maintenance work, especially
+when the cardano-ledger specification changes between eras.
+
+Implications
+------------
+
+* This decision *should not* impact the user facing API of our libraries.
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:134]: insert hyphen for "user-facing"

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+  All of the changes should be internal.
+  Changes to the public facing API should be part of a separate ADR.
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:136]: insert hyphen for "public-facing"

File: doc/adr/0012-commit-to-datatypes-in-cardano-api.rst [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+* Any orphan instances that we currently have in ``plutus-ledger`` will need to
+  be moved to ``cardano-api-extended``.
+
+* The Marconi-related packages will need to work with ``cardano-api`` types
+  instead of ``plutus-ledger-api`` types as Marconi is a full off-chain
+  component.
+  This implies to remove the ``plutus-ledger`` dependency that we currently
```
Comment (joseph-fajen) [inline doc/adr/0012-commit-to-datatypes-in-cardano-api.rst:144]: Suggestion: "This implies removing the"

**Pull request merged**